config/global.php

```diff
@@ -171,40 +171,6 @@
 
         return $ipsResolved;
     },
-    'Zend_Mail_Transport_Abstract' => function () {
-        $mailConfig = Config::getInstance()->mail;
```

Comment:
could this be maybe kept? We're using it on the cloud for example to inject some settings. Haven't looked how phpmailer handles this. Otherwise some DI setting that returns Emails settings could work too. Like 

```php
'mail.settings'  => function () {
   return Config::getInstance()->mail
}
```

We don't really directly want to change the mail config settings using DI on cloud (eg as these values might get saved in the config file etc)

Reply:
Actually, we'll need it to be a class probably as we also use it in WordPress see https://github.com/matomo-org/wp-matomo/blob/1.0.6/config/config.php#L88 and https://github.com/matomo-org/wp-matomo/blob/1.0.6/classes/WpMatomo/Email.php#L39

So we'd maybe need some `Piwik\Mail\Transport` class maybe with a method `send()` or something so we can hook into it and eg ignore PHP Mailer completely

Reply:
Any reason you can't hook into `Mail.send` event and manipulate the settings just before sending the mail? Similar as the tests are capturing the mail, you could also adjust the settings

https://github.com/matomo-org/matomo/blob/2bd03bc2e0028d00cea1dc1d6a36021921b4ff78/plugins/Login/tests/Integration/PasswordResetterTest.php#L197-L214

Reply:
For wordpress you might maybe need to use the `Mail.shouldSend` event, so you are able to abort the normal mail class from sending anything...

Reply:
If listening to shouldSend and returning false it won't trigger `Mail.send` and when listening to Mail.send it will still actually send the mail. Would need to change how this works... through DI is a bit nicer in general as it removes some overhead but not too important. 
We are btw not manipulating the mail we're completely changing the `transport`...

Reply:
Would it be enough to add a `customSend` method to the mail class. e.g.

```
    public function customSend($mimeHeader, $mimeBody)
    {
        Piwik::postEvent('Mail.send.custom', [$mimeHeader, $mimeBody, $this]);
    }
```

That would allow to set `[mail] transport = custom`, causing no mails send with the phpmailer methods, but calling the `customSend` instead.

Reply:
Would need to implement this on Cloud and WordPress to see what it's like. It's maybe not super intuitive in general. Also the `Mail.send` event is not quite clear it indicates you can send the mail there when actually the apps still does it. Could then maybe just change it like this

```
Piwik::postEvent('Mail.send', [$sent= false,$mail]);
if (!$sent) { ... sent mail ... }
```

Would need to test though what happens on cloud there but suppose we could make it work there. Could maybe tweak the above event?